File: prog/experiments/cpp/lexer.cpp
#include "lexer.hpp" // "utils.hpp" -> iostream, string, fstream, vector

using namespace std;

string prepareFile(string file)
{
    file = removeComments(file);
    file = normalize(file);
    file = removeEmptyLines(file);
    return file;
}

string normalize(string file)
{
    string result = "";
    vector<string> lines = strSplit(file, "\n");
    for (string line : lines) {
        string temp = lstrip(line);
        result += temp + "\n";
    }
    return result;
}

string removeComments(string file)
{
    string result = "";
    vector<string> lines = strSplit(file, "\n");
    for (string line : lines) {
        if (line.find("#") != string::npos) {
            size_t comment = line.find("#");
            if (containsEmbeddedString(line)) {
                size_t start = line.find_first_of("\"");
                size_t end = line.find_last_of("\"");
                if (comment > start && comment < end) {
                    result += line + "\n";
                    continue; //! currently cannot place comments after strings that contain '#'
                }
            }
            string temp = line.substr(0, comment);
            result += temp + "\n";
        }
        else {
            result += line + "\n";
        }
    }
    return result;
}

string removeEmptyLines(string file)
{
    vector<string> lines = strSplit(file, "\n");
    string result = "";
    for (string line : lines) {
        if (line.empty()) continue;
        result += line + "\n";
    }
    return result;
}

File: prog/advanced-parser/golang/pkgs/lexer/lexer.go
package lexer

import (
	"strings"
)

func Normalize(file string) string {
	result := ""
	lines := strings.Split(file, "\n")
	for _, line := range lines {
		line = strings.Trim(line, "\t ")
		result += line + "\n"
	}
	return result
}

func RemoveEmptyLines(file string) string {
	result := ""
	lines := strings.Split(file, "\n")
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		result += line + "\n"
	}
	temp := strings.Split(result, "\n")
	temp = temp[:len(temp)-1]
	result = strings.Join(temp, "\n")
	return result
}

func RemoveComments(file string) string {
	result := ""
	lines := strings.Split(file, "\n")
	for _, line := range lines {
		hasComment := false
		quoteCount := 0
		for _, c := range line {
			if c == '"' {
				quoteCount++
			}
			if quoteCount == 1 {
				continue
			}
			if c == '#' {
				hasComment = true
			}
		}
		if hasComment {
			commentIdx := strings.Index(line, "#")
			line = line[:commentIdx]
		}
		result += line + "\n"
	}
	return result
}